### PR TITLE
Add a Taggedlist and use it for InstrListTy in IRFunction

### LINKED ISSUE #1766: Replace llvm::iplist<Instruction> with TaggedList<Instruction> in IRFunction
I want to introduce a new linked list data structure, `TaggedList`. This is an intrusive doubly linked list like `llvm::iplist` with the following differences:

* Each node has a `uint32_t tag` member. These tags are maintained online in a strictly increasing order. This means that the relative position of elements in the list can be determined in (fast) constant time, even while inserting and removing elements from the list.
* The list has a constant time `size()` operation, maintained as a member in the `TaggedList` class. This is needed for the online tag maintenance algorithm.
* Splicing is a linear time operation because the new elements need to be tagged correctly. This is already the case for our specialization of `transferNodesFromList()`.
* The tag also makes it possible to detect when an `end()` iterator is dereferenced or incremented.

I have implemented a prototype of `TaggedList` based on the [Two simplified algorithms for maintaining order in a list](https://scholar.google.com/scholar?cluster=5225046542682967685&hl=en&as_sdt=0,5) paper in order to run benchmarks. List insertions are O(log N) with excellent constant factors.

We already have `class LiveIntervalsInstructionNumbering` in `IROptimizer.cpp`. This change makes that class redundant by providing the same functionality while also allowing online instruction insertions.

The online instruction ordering is needed to implement advanced memory allocation and scheduling in backends.

Preemptively Answered Questions:

* No, we can't extend `llvm::iplist` with this functionality because a) `iplist::size()` is linear time, and b) despite using 27 template classes for configuration and specialization, `iplist` doesn't offer the hooks needed.
* No, we can't implement this as a hack in `IRFunction` because a) we need the constant time `size()` operator, and b) `IRFunction::getInstrs()` exposes the list implementation, and we can't prevent users from modifying the list and invalidating our side tables.

### PR DESCRIPTION
**Pull request by stoklund**
*Description*:
A TaggedList is an intrusive linked list similar to `llvm::ilist`, except a
TaggedList maintains a node ordering so it is possible to compare relative
positions of two nodes in constant time.

*Testing*:
`ninja test`

*Documentation*:
Doxygen comments and unit tests.

Fixes #1766.

### CODE REVIEW

File: include/glow/Base/TaggedList.h [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+    this->addNodeToList(node);
+    impl_.insert(next.node_, node);
+    return iterator(node);
+  }
+
+  /// Remove `node` from this list without deleting it.
+  ///
+  /// Returns a pointer to the removed node.
+  T *remove(T *node) {
+    assert(node && node->inTaggedList());
```
Comment (opti-mix) [inline include/glow/Base/TaggedList.h]: Should it also check that the node belongs to the current list?

File: lib/Base/TaggedList.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  // lo..hi.
+  uint32_t dy = tagCeil - tagFloor;
+  uint32_t dx = count - 1;
+
+  // Make an even larger hold at the insertion point than is strictly
+  // necessary. This anticipates future insertions near the same point.
+  dx += std::min(dx / 8, dy - dx);
+  uint32_t step = dy / dx;
+
+  for (; lo != insertionPoint; lo = lo->nextTaggedNode_, tagFloor += step)
```
Comment (opti-mix) [inline lib/Base/TaggedList.cpp]: Nit: We usually use `{}` in the code-base even for single-line line loop bodies or if-bodies.

**Comment by opti-mix** [comment]
@stoklund Very nice job! 

It is impressive that changing the existing code to use your new `Taggedlist` required just a few changes. I expected more re-factoring ;-)

**Comment by opti-mix** [comment]
@stoklund Have you also tested on bigger NN models using our `run.sh`? Since this change is so intrusive, we should get as much testing as possible.

**Comment by nadavrot** [comment]
The code looks really good. It's clean, documented, simple and tested.

File: include/glow/Base/TaggedList.h [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+  /// Post-decrement.
+  Iterator operator--(int) {
+    Iterator tmp = *this;
+    --*this;
+    return tmp;
+  }
+
+private:
+  // This is either pointing ot a list element of type T or the sentinel
```
Comment (bertmaher) [inline include/glow/Base/TaggedList.h]: typo: "pointing to"

File: include/glow/Base/TaggedList.h (hunk 1 above)
Comment (stoklund) [reply include/glow/Base/TaggedList.h]: That would be nice, but the way the abstraction works, a list node doesn't know which list it belongs to.

We have parent pointers in the `Instruction` nodes, but they are not available to the list implementation itself. See the hacks around `getContainingFunction()`.

File: lib/Base/TaggedList.cpp (hunk 2 above)
Comment (stoklund) [reply lib/Base/TaggedList.cpp]: Thanks, fixed.

**Comment by stoklund** [comment]
> @stoklund Have you also tested on bigger NN models using our `run.sh`? Since this change is so intrusive, we should get as much testing as possible.

Well, it is an intrusive list! ;-)

I've tested the renumbering algorithm with lists up to 256M nodes in length. I'll run it through some real models too.

**Pull request merged**